src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs

```diff
@@ -98,46 +99,81 @@ public ECDsa GetECDsaPrivateKey()
 
+            IntPtr privateKeyPtr;
+
+            // If the certificate has a key handle instead of a key prov info, return the
+            // ephemeral key
+            {
+                int cbData = IntPtr.Size;
+
+                if (Interop.crypt32.CertGetCertificateContextProperty(
+                    certificateContext,
+                    CertContextPropId.CERT_NCRYPT_KEY_HANDLE_PROP_ID,
+                    out privateKeyPtr,
+                    ref cbData))
+                {
+                    handleOptions = CngKeyHandleOpenOptions.EphemeralKey;
+                    return new SafeNCryptKeyHandle(privateKeyPtr, certificateContext);
+                }
+            }
+
             bool freeKey = true;
             SafeNCryptKeyHandle privateKey = null;
+            handleOptions = CngKeyHandleOpenOptions.None;
             try
             {
                 int keySpec = 0;
                 if (!Interop.crypt32.CryptAcquireCertificatePrivateKey(
-                        certificateContext,
-                        CryptAcquireFlags.CRYPT_ACQUIRE_ONLY_NCRYPT_KEY_FLAG,
-                        IntPtr.Zero,
-                        out privateKey,
-                        out keySpec,
-                        out freeKey))
+                    certificateContext,
+                    CryptAcquireFlags.CRYPT_ACQUIRE_ONLY_NCRYPT_KEY_FLAG,
+                    IntPtr.Zero,
+                    out privateKey,
+                    out keySpec,
+                    out freeKey))
                 {
                     int dwErrorCode = Marshal.GetLastWin32Error();
+
+                    // The documentation for CryptAcquireCertificatePrivateKey says that freeKey
+                    // should already be false if "key acquisition fails", and it can be presumed
+                    // that privateKey was set to 0.  But, just in case:
+                    freeKey = false;
+                    privateKey?.SetHandleAsInvalid();
                     return null;
                 }
 
+                // It is very unlikely that Windows will tell us !freeKey other than when reporting failure,
+                // because we set neither CRYPT_ACQUIRE_CACHE_FLAG nor CRYPT_ACQUIRE_USE_PROV_INFO_FLAG, which are
+                // currently the only two success situations documented. However, any !freeKey response means the
+                // key's lifetime is tied to that of the certificate, so re-register the handle as a child handle
+                // of the certificate.
+                if (!freeKey && privateKey != null && !privateKey.IsInvalid)
+                {
+                    var newKeyHandle = new SafeNCryptKeyHandle(privateKey.DangerousGetHandle(), certificateContext);
+                    privateKey.SetHandleAsInvalid();
+                    privateKey = newKeyHandle;
+                    freeKey = true;
+                }
+
                 return privateKey;
             }
-            finally
+            catch
```

Comment:
When adding the copy ctor I noticed this finally block and thought it could be a catch. However, I found in certain cases (IIRC when the context was invalid\freed) the above could fail with error code (no exception) but still fill in the privateKey causing the finally block to bump up the reference count. I assumed this was by design (same in desktop)